Enable ephemeral message, reply with BlockKit UI, and handling Interactive events

ryo-matsuzaki-mdol opened the pull request:
## Sample Image
![Untitled](https://user-images.githubusercontent.com/98291974/167752829-e0b1a1a9-93c5-4895-bfc5-065337d393e9.gif)

### implement `Message#replyEphemeral`
Ruboty applications can reply message as [ephemeral message](https://api.slack.com/methods/chat.postEphemeral).

### `Message#reply` `Message#replyEphemeral` can receive `block` argument
These methods can handle [BlockKit](https://api.slack.com/block-kit) JSON as argument: `block`.

### (experimental) Interactive events of BlockKit can be passed to handlers
This is experimental feature. 
BlockKit's interactive events is handled as Ruboty-command: `slack_interactive::#{action_id}`.
Ruboty handlers can receive them by catching above command.

### (experimental) implement `Message#delete` for ephemeral message interaction
This is experimental feature. 
Ruboty message made from BlockKit's interactive events can use `delete` method.
This is useful to clean up ephemeral message containing BlockKit UI.
It can be used only for ephemeral message.

ryo-matsuzaki-mdol commented on ruboty-slack_socket_mode.gemspec:
```diff
@@ -1,16 +1,16 @@
+  spec.name          = 'ruboty-slack_socket_mode'
+  spec.version       = Ruboty::SlackSocketMode::VERSION
+  spec.authors       = ['Repro Inc.']
+  spec.email         = ['support@repro.io']
```
[self comment] TODO: `spec.email` に記載するメールアドレスとしてこれが妥当かどうか確認する。

ryo-matsuzaki-mdol replied:
resolved ✅

ryo-matsuzaki-mdol commented on lib/ruboty/adapters/slack_socket_mode.rb:
```diff
@@ -86,16 +87,18 @@ def init
         realtime.on_text do |data|
-          method_name = "on_#{data['type']}".to_sym
-          send(method_name, data) if respond_to?(method_name, true)
+          next if data['type'] != 'events_api'
```
https://api.slack.com/apis/connections/socket-implement#events
SocketModeでの payload 形式については上のページを参考にして実装しております！
ただ、 `type` の一覧は記載されていません・・

ドキュメントにある例で網羅されていると信じるなら `events_api` `interactive` `slash_commands` の3値を取りそうではあるのですが。

takaishi replied:
なるほど…！

takaishi replied:
このPRではevents_apiにのみ対応すればいいはずなので、これで問題はなさそう 👍

takaishi commented on ruboty-slack_socket_mode.gemspec:
```diff
@@ -1,16 +1,16 @@
+  spec.name          = 'ruboty-slack_socket_mode'
+  spec.version       = Ruboty::SlackSocketMode::VERSION
+  spec.authors       = ['Repro Inc.']
+  spec.email         = ['support@repro.io']
```
一旦 `ryo.takaishi@repro.io` にしておいてもらってもよいでしょうか

ryo-matsuzaki-mdol replied:
確認感謝です！変更いたしました〜

ryo-matsuzaki-mdol commented on lib/ruboty/adapters/slack_socket_mode.rb:
```diff
@@ -111,19 +114,23 @@ def connect
+        @client ||= ::Slack::Web::Client.new(token: ENV['SLACK_BOT_TOKEN'])
       end
 
       def realtime
```
WebSocketのClient・・ということで `socket` という命名にしてみました！

ryo-matsuzaki-mdol commented:
@takaishi いただいたコメント全体について対応してみました！
改めてご確認いただけますと幸いです！

takaishi commented:
@okkez こちら、追加でレビューお願いできますか？

takaishi commented on lib/ruboty/slack_socket_mode/client.rb:
```diff
@@ -25,7 +25,11 @@ def on_text(&block)
-            block.call(JSON.parse(message.data))
+            data = JSON.parse(message.data)
+
+            # ACK response
```
[nits] SocketModeで必要である旨のコメントや、ドキュメントへのリンクがあるとなおよさそうです！

ryo-matsuzaki-mdol replied:
最小限ではありますが、リンクを添えてコメント追記してみました！

okkez commented:
大半を https://github.com/rosylilly/ruboty-slack_rtm/ から流用しているとのことでしたが fork していないのは何故でしょうか？
個人的には GitHub 上で fork 元がちゃんとわかるようにしてから始めなければならないと考えています。
最初から public repo でやるならなおさら。

ryo-matsuzaki-mdol commented:
> 大半を https://github.com/rosylilly/ruboty-slack_rtm/ から流用しているとのことでしたが fork していないのは何故でしょうか？

ruboty-slack_rtm とは別物を作る前提で、本家へのPRや本家からの取り込みを行うことが必要ないので clone から始めていました。最近見た例(https://github.com/suzuki-shunsuke/tfcmt) もあり、それを模倣したというのもあります・・

ただ、OSSとしてコードを作っていく経験があまりなく、それらが fork しない理由として正しいのか？はあまり自信なく進めてしまったと思っております。類似のケースで fork している例もありました。

okkez さんのコメント通り、public repo であって、どちらも GitHub 上にホストされるリポジトリであれば明示的に fork 元がわかるようになっているのが一番良い形なような気がしました。 fork として再作成してみようかと思うのですが、いきなりリポジトリが消えたりしていると混乱を生みそうなので、 @takaishi さんとも同期が取れるタイミングで話をしてみようかと思います！

ryo-matsuzaki-mdol committed:
implement minimal slack interactive features

- Message.reply with :blocks can post interactive UIs
- implement Message.reply_ephemeral
- implement Message.delete for ephemeral message

ryo-matsuzaki-mdol committed:
fix interactive data digging

- use dig method instead of multi-array access
- make property to refer env: RUBOTY_NAME
